octoprint_mrbeam/comm_acc2.py

```diff
@@ -208,36 +241,50 @@ def _send_loop(self):
 				self._send_event.wait(1)
 				self._send_event.clear()
 			except:
-				self._logger.exception("Something crashed inside the sending loop, please report this to Mr. Beam")
+				self._logger.exception("Something crashed inside the sending loop, please report this to Mr Beam.")
 				errorMsg = "See octoprint.log for details"
 				self._log(errorMsg)
 				self._errorValue = errorMsg
 				self._changeState(self.STATE_ERROR)
 				eventManager().fire(OctoPrintEvents.ERROR, {"error": self.getErrorString()})
 				self._logger.dump_terminal_buffer(level=logging.ERROR)
 
-	def _metric_loop(self):
-		self._metricf = open('metric.tmp','w')
-		self._metricf.write("1 sec interval")
-		while self._metric_active:
-			time.sleep(1)
-			if self._metric_time is not None:
-				t = time.time()
-				#s = "Metric: %f [chars/sec]" % (self._metric_chars / (t - self._metric_time))
-				s = "%.2f" % (self._metric_chars / (t - self._metric_time))
-				self._metric_time = None
-				self._metric_chars = 0
-				self._metricf.write(s)
-				#self._log(s)
-		self._metricf.close()
-
 	def _sendCommand(self, cmd=None):
 		if cmd is None:
 			if self._cmd is None and self._commandQueue.empty():
 				return
 			elif self._cmd is None:
 				self._cmd = self._commandQueue.get()
-			if sum([len(x) for x in self._acc_line_buffer]) + len(self._cmd) +1 < self.RX_BUFFER_SIZE-5:
+			if self._cmd == self.COMMAND_FLUSH:
+				# FLUSH waits until we're no longer waiting for any OKs from GRBL
+				if self._sync_command_ts <=0:
+					self._sync_command_ts = time.time()
+				if len(self._acc_line_buffer) <= 0:
+					self._cmd = None
+					self._log("FLUSHed ({}ms)".format(int(1000*(time.time() - self._sync_command_ts))))
+					self._sync_command_ts = -1
+				self._send_event.set()
+			elif self._cmd == self.COMMAND_SYNC:
+				# SYNC waits until we're no longer waiting for any OKs from GRBL and GRBL has reported to no be busy anymore.
+				# Still experimential: We need to test/verify/implement:
+				# - Maybe we need to turn off the laser here...
+				# - What if RX buffer remains 1 and never becomes 0? sync should handle/correct this
+				# - Do we need a timeout or something?
+				# self._logger.info("ANDYTEST SYNC: buffer: %s grbl_state: %s, grbl_rx: %s", len(self._acc_line_buffer), self._grbl_state, self._grbl_rx_status )
+				if self._sync_command_ts <=0:
+					self._sync_command_ts = time.time()
+				if len(self._acc_line_buffer) <= 0 and not self._grbl_state in self.GRBL_SYNC_COMMAND_WAIT_STATES:
```

Comment:
Add log when SYNC is handled for the first time with:
- Add grbl's RX value and acc_line_buffer char len for debugging/investigation

Reply:
done